### PR TITLE
coredevice: mirny/adf5355: add basic high-level interface (WIP)

### PR DESCRIPTION
**Pull request by airwoodix**
# ARTIQ Pull Request

## Description of Changes

This patch adds a basic high-level interface for setting a fixed frequency output from the Mirny ADF5355 PLL based synthesizers.  Algorithms are largely inspired by the ones used in the [linux driver](https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/frequency/adf5355.c).

After revision, I have the following leftover open points / remarks:
  - no support for `clk_div` on v1.0 boards: sounds like one doesn't need that, v1.0 boards are outdated and do not support extensions boards like [Almazny](https://github.com/sinara-hw/Almazny);
  - the register map is automatically generated using [regdesc](https://github.com/airwoodix/regdesc). Where should the generation code go? A unified mechanism to generate register maps would be great (also applicable to chips on Urukul and Phaser for example) but seems to be out of the scope of this PR;
  - testing is still pretty weak. It'd be great if other Mirny users would provide feedback;
  - release notes entry is part of ARTIQ 6 release;

Basic testing was performed based on the following snippet:

```python
from artiq.experiment import *
from pprint import pprint

class TestMirny(EnvExperiment):
    def build(self):
        self.setattr_device("core")
        self.setattr_device("mirny0_ch0")

        self.ch = self.mirny0_ch0

    @kernel
    def run(self):
        self.core.reset()

        self.ch.cpld.init()
        self.ch.init()

        self.ch.set_att_mu(160)
        self.ch.sw.on()
        self.core.break_realtime()

        self.ch.set_frequency(1000*MHz)
        delay(5*ms)
        print(self.ch.read_muxout())

    def analyze(self):
        pprint(self.ch.info())
```

I tested discrete frequencies in the range 60 MHz to 1.4 GHz on a spectrum analyzer. Lacking a counter, I cannot assess the fine frequency accuracy. Lock time seems to be around 5 ms, judging from the necessary delay to read MUXOUT high.

Comments most welcome.
Thanks.

### Related Issue

#1421 

## Type of Changes

|   | Type |
| ------------- | ------------- |
|   | :bug: Bug fix  |
| ✓  | :sparkles: New feature |
|   | :hammer: Refactoring  |
|   | :scroll: Docs |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Update [RELEASE_NOTES.md](../RELEASE_NOTES.md) if there are noteworthy changes, especially if there are changes to existing APIs.
- [x] Close/update issues.
- [x] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.
- [x] Add and check docstrings and comments
- [x] Check, test, and update the [unittests in /artiq/test/](../artiq/test/) or [gateware simulations in /artiq/gateware/test](../artiq/gateware/test)

### Documentation Changes

- [x] Check, test, and update the documentation in [doc/](../doc/). Build documentation (`cd doc/manual/; make html`) to ensure no errors.

### Git Logistics

- [ ] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

**Comment by sbourdeauducq** [comment]
>     * lacking `gcd` support on the coredevice, `MOD2` is set to its maximum (so minimum channel spacing)

This should compile (after rewriting the tuple unpacking):
https://scipython.com/book/chapter-2-the-core-python-language-i/examples/euclids-algorithm-for-finding-the-gcd-of-a-number/

File: artiq/coredevice/adf5355.py [hunk 1]
```diff
@@ -53,13 +118,393 @@ def read_muxout(self):
...
+            # RELOCK AT WANTED PFD FREQUENCY
+
+            for i in [4, 2, 1]:
+                self.write(self.regs[i])
+
+            # force-disable autocal
+            self.write(self.regs[0] & ~ADF5355_REG0_AUTOCAL(1))
+
+    @portable
+    def f_pfd(self) -> TInt64:
```
Comment (sbourdeauducq) [inline artiq/coredevice/adf5355.py:244]: Is the type annotation required here? (OK to leave it, just curious)

File: artiq/coredevice/adf5355.py [hunk 2]
```diff
@@ -53,13 +118,393 @@ def read_muxout(self):
...
+        ) | ADF5355_REG2_AUX_MOD_LSB_VALUE_GET(self.regs[2])
+
+    @portable
+    def ref_counter(self) -> TInt32:
+        """
+        Return the reference counter value (R) for the cached set of registers.
+        """
+        return ADF5355_REG4_R_COUNTER_GET(self.regs[4])
+
+    @rpc
```
Comment (sbourdeauducq) [inline artiq/coredevice/adf5355.py:306]: rpc is the default

File: artiq/coredevice/adf5355.py (hunk 1 above)
Comment (airwoodix) [reply artiq/coredevice/adf5355.py]: It's indeed not required (quick test).

**Comment by airwoodix** [comment]
Thanks for the feedback!

The last commits add:

-  the missing more general PLL parameters calculation
- a naming move to ADF5356, matching the dominant? chip on Mirny. The driver should work with ADF435[56] variants but I didn't test
- output A enable/disable and output power programming
- unit tests and documentation fixes

I also edited the top message with some more minor open issues.

**Pull request merged**